crates/vsock/src/vhu_vsock_thread.rs

```diff
@@ -462,76 +462,81 @@ impl VhostUserVsockThread {
 
         let queue = vring_mut.get_queue_mut();
 
-        while let Some(mut avail_desc) = queue
-            .iter(atomic_mem.memory())
-            .map_err(|_| Error::IterateQueue)?
-            .next()
-        {
-            used_any = true;
-            let mem = atomic_mem.clone().memory();
-
-            let head_idx = avail_desc.head_index();
-            let used_len = match VsockPacket::from_rx_virtq_chain(
-                mem.deref(),
-                &mut avail_desc,
-                self.tx_buffer_size,
-            ) {
-                Ok(mut pkt) => {
-                    let recv_result = match rx_queue_type {
-                        RxQueueType::Standard => self.thread_backend.recv_pkt(&mut pkt),
-                        RxQueueType::RawPkts => self.thread_backend.recv_raw_pkt(&mut pkt),
-                    };
-
-                    if recv_result.is_ok() {
-                        PKT_HEADER_SIZE + pkt.len() as usize
-                    } else {
-                        queue.iter(mem).unwrap().go_to_previous_position();
-                        break;
+        let mut iter_has_elemnt = true;
+        while iter_has_elemnt {
+            let queue_iter = queue
+                .iter(atomic_mem.memory())
+                .map_err(|_| Error::IterateQueue)?;
+
+            iter_has_elemnt = false;
+            for mut avail_desc in queue_iter {
+                used_any = true;
+                iter_has_elemnt = true;
+                let mem = atomic_mem.clone().memory();
+
+                let head_idx = avail_desc.head_index();
+                let used_len = match VsockPacket::from_rx_virtq_chain(
+                    mem.deref(),
+                    &mut avail_desc,
+                    self.tx_buffer_size,
+                ) {
+                    Ok(mut pkt) => {
+                        let recv_result = match rx_queue_type {
+                            RxQueueType::Standard => self.thread_backend.recv_pkt(&mut pkt),
+                            RxQueueType::RawPkts => self.thread_backend.recv_raw_pkt(&mut pkt),
+                        };
+
+                        if recv_result.is_ok() {
+                            PKT_HEADER_SIZE + pkt.len() as usize
+                        } else {
+                            queue.iter(mem).unwrap().go_to_previous_position();
+                            break;
```

Comment:
Now that we have 2 loops nested, IIUC this should exit both loops, so we may need to replace it with a `return Ok(used_any)`, or set `iter_has_elemnt` to false.

If you prefer the last one, maybe then better to change the name of that variable to something like `processing_rx`.